src/share/routes.js

```diff
@@ -303,6 +331,37 @@ function register(router) {
 
         res.send(note.getContent());
     });
+
+    // Used for searching, require noteId so we know the subTreeRoot
+    router.get('/share/api/search/:noteId', (req, res, next) => {
+        shacaLoader.ensureLoad();
+
+        let note;
+
+        if (!(note = checkNoteAccess(req.params.noteId, req, res))) {
+            return;
+        }
+
+        const {query} = req.query;
+
+        if (!query?.trim()) {
+            return res.status(400).json({ message: "'query' parameter is mandatory." });
+        }
+
+        const subRootPath = getSharedSubTreeRoot(note);
+        const subRoot = subRootPath.note;
+        const searchContext = new SearchContext({ancestorNoteId: subRoot.noteId});
+        const searchResults = searchService.findResultsWithQuery(query, searchContext);
```

Comment:
TBH I was quite uncomfortable with this reusing of "core" Trilium logic.

Sharing functionality is intentionally kept very separated from the main codebase and especially from the data (becca) to minimize the risk of bugs in the main codebase leaking data to the public internet. This basically means that all non-trivial logic is duplicated in the share codebase and this would normally need to happen for search as well.

But since I'm not keen on writing share-only search functionality, the second look doesn't seem so bad - we're taking only the `noteId` from the search and then data retrieval still happens only through `shaca` (which contains only shared notes). Thus even with buggy search code, there shouldn't be any leak of non-shared data.

Reply:
I did think about the exposure risk of using the general search functionality but as you said since everything is filtered through `shaca` there should be no results passed back to the client-side JS that are not already shared. Plus using the `SearchContext` should already limit our results fairly well.